Add income and spending insight cards to dashboard

JoshAntBrown opened the pull request:
This PR implements the income and spending insight cards for the dashboard.

<img width="1624" alt="Screenshot 2024-04-23 at 14 50 25" src="https://github.com/maybe-finance/maybe/assets/1793797/f00e873e-54ef-4410-8982-9a19ae4c77ab">

## Requirements
* [x]  Should generate a time series of income and spending for a `Family`, scoped to `Period.last_30_days`
  * [x]  Each data point in the series represents a "Trailing 30 day sum" of income (negative transaction values)
  * [x]  The "headline" number shows the last value in the series
  * [x]  The "trend" value is the overall time series trend
* [x]  Should find the top 3 accounts with the _most_ income or spending within the last 30 days and display the sum
  * [x]  Each account is an anchor that goes to that account page
  * [x]  Bank icons are not in-scope (just use a generic letter as shown in designs)
  * [x]  A "more" counter should show if there are more than 3 accounts
* [x]  Display a trendline of series.  No tooltips, no axes.
* [x]  These insights should be included in the dashboard


/claim #645
Resolves #645

JoshAntBrown commented:
/claim #645

zachgoll commented on app/models/family.rb:
```diff
@@ -28,6 +28,78 @@ def snapshot(period = Period.all)
+      income_series: TimeSeries.new(income, favorable_direction: "up"),
+      spending_series: TimeSeries.new(spending, favorable_direction: "down")
+    }
+  end
```
I like the overall direction of this query.  Definitely a complex one (that will likely need to be modified as we start adding various transaction types), so I'm thinking we may need a few test cases for documentation purposes.  As a quick summary of this comment:

- Series length should always be 30 days, regardless of how many currencies a `Family` has
- We'll need to handle exchange rate conversions for this rollup
- Consider using a CTE for clarity of the query? - Rails 7 has the `with` method that _may_ make this a little easier to scan top-to-bottom?

Looking at the output for the demo account (`rake demo_data:reset`):

![CleanShot 2024-04-23 at 10 53 17](https://github.com/maybe-finance/maybe/assets/16676157/e31017a9-b8f7-4a82-878a-93a86abdd95c)

In the screenshot above, I would expect these insights to _always_ show in the `Family.currency` (the demo account is `USD`, but one of the top moving accounts is `EUR`).  Also, it appears that both currency series from the query are being included, causing the count to be `62` rather than the intended `31`.

To handle multiple currencies, I'm thinking we'll need an _additional_ step that will normalize everything into the `Family.currency` prior to summing it all up
  - Since we cannot _depend on_ exchange rates being available, we'll either need to _omit_ unconverted amounts from the aggregation or omit accounts that contain "foreign" currencies from the calculation entirely.
  - My guess is that it would be easiest to perform a join on `exchange_rates` table somewhere within this query and do the filtering there (rather than trying to call `exchange_to` on each money instance after-the-fact).  In this strategy, we'd basically be saying, "we'll try to convert, but if the app doesn't already have the needed exchange rates cached to the DB, we can't calculate this and will omit it"
  - We could still technically show a "top mover" in a native foreign currency (since we have a series for all currencies), but series that cannot be converted via the query should never make it into that headline number

I've been using [this spreadsheet](https://docs.google.com/spreadsheets/d/18LN5N-VLq4b49Mq1fNwF7_eBiHSQB46qQduRtdAEN98/edit#gid=0) to visualize the expected results and conversions a little bit better.  Just throwing it out there in case it helps with this one.

As this is a tricky one, I'm open to ideas!  All the comments above are just my best guess at how we could approach it.

----------

Here's a _sketch_ of what I was thinking for the exchange rates (haven't tested thoroughly, using some hardcoded placeholders):

```sql
WITH date_series AS (
	SELECT * 
	FROM 
		generate_series('2024-02-23'::date, '2024-04-23'::date, interval '1 day') AS gs(date) 
		CROSS JOIN (SELECT DISTINCT currency FROM transactions) as c
), normalized AS (
	SELECT
        gs.date,
        gs.currency,
        COALESCE(SUM(t.amount) FILTER (WHERE t.amount > 0), 0) AS spending,
        COALESCE(SUM(-t.amount) FILTER (WHERE t.amount < 0), 0) AS income
    FROM
        date_series gs
        LEFT JOIN transactions t ON t.date = gs.date AND t.currency = gs.currency
    GROUP BY gs.date, gs.currency
), rolling AS (
	SELECT 
		nq.*,
		SUM(spending) OVER (PARTITION BY currency ORDER BY date RANGE BETWEEN INTERVAL '30 days' PRECEDING and CURRENT ROW) as rolling_spend,
		SUM(income) OVER (PARTITION BY currency ORDER BY date RANGE BETWEEN INTERVAL '30 days' PRECEDING and CURRENT ROW) as rolling_income
	FROM normalized nq
	ORDER BY nq.date
), converted_rolling AS (
	SELECT
        r.date,
        COALESCE(r.rolling_spend * er.rate, 0) AS converted_rolling_spend,
        COALESCE(r.rolling_income * er.rate, 0) AS converted_rolling_income
    FROM
        rolling r
        LEFT JOIN exchange_rates er ON r.currency = er.base_currency AND er.converted_currency = 'USD'
    GROUP BY r.date, er.rate, r.rolling_spend, r.rolling_income
) SELECT * FROM converted_rolling;
```

IMO, these CTEs read a little easier but I know it completely deviates from ActiveRecord convention.  Not sure what the best approach is?

JoshAntBrown replied:
Thanks! Yeah it's been a tricky one, that makes sense regarding the currencies I was mirroring what was going on with the net worth chart but this makes more sense.

I'll take a look at working the currency conversion into things, and while I do that I'll see if I can do anything to improve the readability.

zachgoll replied:
Sounds good.  Yeah, the NW chart benefits from having the `account_balances` table to read from which is "normalized" (which includes currency conversion) during the account sync process, which makes it easier to just filter by the family's currency (and assume any account balances that aren't in this subset couldn't be converted).

This one is definitely tricky since we have to normalize and convert on-the-fly

zachgoll replied:
@JoshAntBrown I had previously experimented around with a DB table called `family_snapshots` that would essentially act as a daily "cache" table for these sorts of calculations (similar to `account_balances`):

```
date, net_worth, assets, liabilities, rolling_30_day_income, rolling_30_day_expense
```

We'd essentially be doing the same calculations here but rather than on-the-fly we'd have a daily cron running to calculate them and then here we'd just read the values.

Not sure if that would really make any of this easier but thought I'd float the idea in case it helped with this one at all.

zachgoll reviewed (commented):
The structure of this looks good to me, just added some notes regarding tests / exchange rate conversions for handling multi-currency.

JoshAntBrown commented on app/models/family.rb:
```diff
@@ -28,6 +28,77 @@ def snapshot(period = Period.all)
+
+  def snapshot_transactions
+    period = Period.last_30_days
+    days_rolling = 30
```
Yep makes sense to me, we can just count the days in period we're using and then use that to build the query. I'll get that updated.

JoshAntBrown commented:
I've had a bit of a refactor and added a test in a similar fashion to the other account snapshot.

Hopefully this is a bit more manageable now without straying too far from Active Record conventions.

Here's my copy of the test cases, updated to include spending and income on a rolling basis:
https://docs.google.com/spreadsheets/d/112tRiy0l8WWngEGMdEX2JprYJmYWUW--B2UDUiW9nOE/edit

zachgoll reviewed (approved):
Looks great, thanks for adding those test cases!  I have copy/pasted the expected results into that main spreadsheet for now and we'll eventually get all those into an easier format to collaborate on.
